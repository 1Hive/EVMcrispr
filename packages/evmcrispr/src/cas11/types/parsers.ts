import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
