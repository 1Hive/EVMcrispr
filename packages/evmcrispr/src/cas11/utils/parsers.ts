import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
