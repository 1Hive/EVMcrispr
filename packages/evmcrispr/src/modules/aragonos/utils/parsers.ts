import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
