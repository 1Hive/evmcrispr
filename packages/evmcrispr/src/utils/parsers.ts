import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
