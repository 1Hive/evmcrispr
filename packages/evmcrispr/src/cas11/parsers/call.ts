import type { Parser } from 'arcsecond';
import {
  char,
  choice,
  coroutine,
  letters,
  possibly,
  recursiveParser,
} from 'arcsecond';

import type { StringLiteralNode } from '../types';
import { NodeType } from '../types';

import {
  addressParser,
  probableIdentifierParser,
  variableIdentifierParser,
} from './primaries';
import { argumentsParser } from './expression';
import { helperFunctionParser } from './helper';

export const callExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    coroutine(function* () {
      const target = yield choice([
        addressParser,
        variableIdentifierParser,
        helperFunctionParser,
        probableIdentifierParser,
      ]);

      let res = yield char(':');
      let callExpressionNode: any = target;

      do {
        const callee = yield letters.map(
          (value): StringLiteralNode => ({
            type: NodeType.StringLiteral,
            value,
          }),
        );

        const args = yield argumentsParser;

        callExpressionNode = {
          type: NodeType.CallExpression,
          target: callExpressionNode,
          callee,
          args,
        };

        res = yield possibly(char(':'));
      } while (res);

      return callExpressionNode;
    }),
);
