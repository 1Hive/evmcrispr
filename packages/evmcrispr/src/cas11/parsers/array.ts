import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
