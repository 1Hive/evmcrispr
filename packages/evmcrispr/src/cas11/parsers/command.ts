import type { Parser } from 'arcsecond';
import {
  choice,
  coroutine,
  endOfInput,
  many1,
  recursiveParser,
  regex,
} from 'arcsecond';

import type { CommandExpressionNode, CommandIdentifierNode } from '../types';
import { NodeType } from '../types';

import { expressionParser } from './expression';
import { endOfLine, optionalWhitespace, whitespace } from './utils';

const commandNameRegex =
  /^(?:(?<module>[a-zA-Z-]{1,63}(?<!-))(?::))?(?<command>[a-zA-Z-]{1,63}(?<!-))/;
const commandNameParser = regex(commandNameRegex).map(
  (value): CommandIdentifierNode => {
    const res = commandNameRegex.exec(value);
    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
    const { module, command } = res?.groups!;

    const c: CommandIdentifierNode = {
      type: NodeType.CommandIdentifier,
      value: command,
    };

    c.value = command;
    if (module) c.module = module;

    return c;
  },
);

const commandArgsParser = many1(
  coroutine(function* () {
    yield whitespace;

    const args = yield expressionParser;

    return args;
  }),
);

export const endOfCommandParser = choice([endOfLine, endOfInput]);

export const commandExpressionParser: Parser<
  CommandExpressionNode,
  string,
  any
> = recursiveParser(() =>
  coroutine(function* () {
    yield optionalWhitespace;

    const name = yield commandNameParser;

    const args = yield commandArgsParser;

    yield optionalWhitespace;

    yield endOfCommandParser;

    return {
      type: NodeType.CommandExpression,
      name,
      args,
    } as unknown as CommandExpressionNode;
  }),
);
