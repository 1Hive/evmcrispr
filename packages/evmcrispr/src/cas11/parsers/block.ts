import type { Parser } from 'arcsecond';
import { char, coroutine, recursiveParser } from 'arcsecond';

import type { BlockExpressionNode } from '../types';
import { NodeType } from '../types';
import { commandExpressionParser } from './command';
import {
  endOfLine,
  optionalEmptyLines,
  optionalWhitespace,
  surroundedBy,
} from './utils';

export const blockExpressionParser: Parser<BlockExpressionNode, string, any> =
  recursiveParser(() =>
    coroutine(function* () {
      yield char('(');

      yield optionalWhitespace;

      yield endOfLine;

      const scopedCommands = yield optionalEmptyLines(commandExpressionParser);

      yield surroundedBy(optionalWhitespace)(char(')'));

      return {
        type: NodeType.BlockExpression,
        body: scopedCommands,
      } as unknown as BlockExpressionNode;
    }),
  );
