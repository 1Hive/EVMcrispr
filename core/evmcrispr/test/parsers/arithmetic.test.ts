import { runParser } from '@1hive/evmcrispr-test-common';
import type { Case } from '@1hive/evmcrispr-test-common';

import { arithmeticParser } from '../../src/parsers/arithmetic';

describe.concurrent('Parsers - arithmetic', () => {
  describe.each<Case>([
    { title: '', value: '(9 + 5 - 4 * 4 / 3 ^ 2)' },
    {
      title: 'with trailing spaces at both ends',
      value: '(   9 + 5 - 4 * (4 / 3) ^ 3    )',
    },
    {
      title: 'with in-between trailing spaces',
      value: '(9 +    5    - 4    *     4   /    3)',
    },
    {
      title: 'containing priority parenthenses',
      value: '(9^33 + (5 - 4) * (4 / 3))',
    },
    {
      title: 'containing helper functions and call expressions',
      value:
        '(90.45e18 + (5000e18 - @token.balance( DAI, @me)) * (someContract::getAmount() / 3) + $some-Variable ^ 2)',
    },
  ])('', ({ title, value }) => {
    it(`should parse an arithmetic operation ${title} correctly`, () => {
      const parsedValue = runParser(arithmeticParser, value);

      expect(parsedValue).toMatchSnapshot();
    });
  });
});
