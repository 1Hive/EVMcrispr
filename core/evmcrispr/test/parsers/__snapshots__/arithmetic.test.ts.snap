// Vitest Snapshot v1

exports[`Parsers - arithmetic > should parse an arithmetic operation  correctly 1`] = `
{
  "left": {
    "left": {
      "loc": {
        "end": {
          "col": 2,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "9",
    },
    "loc": {
      "end": {
        "col": 6,
        "line": 1,
      },
      "start": {
        "col": 1,
        "line": 1,
      },
    },
    "operator": "+",
    "right": {
      "loc": {
        "end": {
          "col": 6,
          "line": 1,
        },
        "start": {
          "col": 5,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "5",
    },
    "type": "BinaryExpression",
  },
  "loc": {
    "end": {
      "col": 22,
      "line": 1,
    },
    "start": {
      "col": 1,
      "line": 1,
    },
  },
  "operator": "-",
  "right": {
    "left": {
      "left": {
        "loc": {
          "end": {
            "col": 10,
            "line": 1,
          },
          "start": {
            "col": 9,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "loc": {
        "end": {
          "col": 14,
          "line": 1,
        },
        "start": {
          "col": 9,
          "line": 1,
        },
      },
      "operator": "*",
      "right": {
        "loc": {
          "end": {
            "col": 14,
            "line": 1,
          },
          "start": {
            "col": 13,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "type": "BinaryExpression",
    },
    "loc": {
      "end": {
        "col": 22,
        "line": 1,
      },
      "start": {
        "col": 9,
        "line": 1,
      },
    },
    "operator": "/",
    "right": {
      "left": {
        "loc": {
          "end": {
            "col": 18,
            "line": 1,
          },
          "start": {
            "col": 17,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "3",
      },
      "loc": {
        "end": {
          "col": 22,
          "line": 1,
        },
        "start": {
          "col": 17,
          "line": 1,
        },
      },
      "operator": "^",
      "right": {
        "loc": {
          "end": {
            "col": 22,
            "line": 1,
          },
          "start": {
            "col": 21,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`Parsers - arithmetic > should parse an arithmetic operation containing helper functions and call expressions correctly 1`] = `
{
  "left": {
    "left": {
      "loc": {
        "end": {
          "col": 9,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "power": 18,
      "type": "NumberLiteral",
      "value": "90.45",
    },
    "loc": {
      "end": {
        "col": 82,
        "line": 1,
      },
      "start": {
        "col": 1,
        "line": 1,
      },
    },
    "operator": "+",
    "right": {
      "left": {
        "left": {
          "loc": {
            "end": {
              "col": 20,
              "line": 1,
            },
            "start": {
              "col": 13,
              "line": 1,
            },
          },
          "power": 18,
          "type": "NumberLiteral",
          "value": "5000",
        },
        "loc": {
          "end": {
            "col": 48,
            "line": 1,
          },
          "start": {
            "col": 13,
            "line": 1,
          },
        },
        "operator": "-",
        "right": {
          "args": [
            {
              "loc": {
                "end": {
                  "col": 42,
                  "line": 1,
                },
                "start": {
                  "col": 39,
                  "line": 1,
                },
              },
              "type": "ProbableIdentifier",
              "value": "DAI",
            },
            {
              "args": [],
              "loc": {
                "end": {
                  "col": 47,
                  "line": 1,
                },
                "start": {
                  "col": 44,
                  "line": 1,
                },
              },
              "name": "me",
              "type": "HelperFunctionExpression",
            },
          ],
          "loc": {
            "end": {
              "col": 48,
              "line": 1,
            },
            "start": {
              "col": 23,
              "line": 1,
            },
          },
          "name": "token.balance",
          "type": "HelperFunctionExpression",
        },
        "type": "BinaryExpression",
      },
      "loc": {
        "end": {
          "col": 82,
          "line": 1,
        },
        "start": {
          "col": 12,
          "line": 1,
        },
      },
      "operator": "*",
      "right": {
        "left": {
          "args": [],
          "loc": {
            "end": {
              "col": 78,
              "line": 1,
            },
            "start": {
              "col": 53,
              "line": 1,
            },
          },
          "method": "getAmount",
          "target": {
            "loc": {
              "end": {
                "col": 65,
                "line": 1,
              },
              "start": {
                "col": 53,
                "line": 1,
              },
            },
            "type": "ProbableIdentifier",
            "value": "someContract",
          },
          "type": "CallExpression",
        },
        "loc": {
          "end": {
            "col": 82,
            "line": 1,
          },
          "start": {
            "col": 53,
            "line": 1,
          },
        },
        "operator": "/",
        "right": {
          "loc": {
            "end": {
              "col": 82,
              "line": 1,
            },
            "start": {
              "col": 81,
              "line": 1,
            },
          },
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "loc": {
    "end": {
      "col": 104,
      "line": 1,
    },
    "start": {
      "col": 1,
      "line": 1,
    },
  },
  "operator": "+",
  "right": {
    "left": {
      "loc": {
        "end": {
          "col": 100,
          "line": 1,
        },
        "start": {
          "col": 86,
          "line": 1,
        },
      },
      "type": "VariableIdentifier",
      "value": "$some-Variable",
    },
    "loc": {
      "end": {
        "col": 104,
        "line": 1,
      },
      "start": {
        "col": 86,
        "line": 1,
      },
    },
    "operator": "^",
    "right": {
      "loc": {
        "end": {
          "col": 104,
          "line": 1,
        },
        "start": {
          "col": 103,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "2",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`Parsers - arithmetic > should parse an arithmetic operation containing priority parenthenses correctly 1`] = `
{
  "left": {
    "left": {
      "loc": {
        "end": {
          "col": 2,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "9",
    },
    "loc": {
      "end": {
        "col": 5,
        "line": 1,
      },
      "start": {
        "col": 1,
        "line": 1,
      },
    },
    "operator": "^",
    "right": {
      "loc": {
        "end": {
          "col": 5,
          "line": 1,
        },
        "start": {
          "col": 3,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "33",
    },
    "type": "BinaryExpression",
  },
  "loc": {
    "end": {
      "col": 24,
      "line": 1,
    },
    "start": {
      "col": 1,
      "line": 1,
    },
  },
  "operator": "+",
  "right": {
    "left": {
      "left": {
        "loc": {
          "end": {
            "col": 10,
            "line": 1,
          },
          "start": {
            "col": 9,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "5",
      },
      "loc": {
        "end": {
          "col": 14,
          "line": 1,
        },
        "start": {
          "col": 9,
          "line": 1,
        },
      },
      "operator": "-",
      "right": {
        "loc": {
          "end": {
            "col": 14,
            "line": 1,
          },
          "start": {
            "col": 13,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "type": "BinaryExpression",
    },
    "loc": {
      "end": {
        "col": 24,
        "line": 1,
      },
      "start": {
        "col": 8,
        "line": 1,
      },
    },
    "operator": "*",
    "right": {
      "left": {
        "loc": {
          "end": {
            "col": 20,
            "line": 1,
          },
          "start": {
            "col": 19,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "loc": {
        "end": {
          "col": 24,
          "line": 1,
        },
        "start": {
          "col": 19,
          "line": 1,
        },
      },
      "operator": "/",
      "right": {
        "loc": {
          "end": {
            "col": 24,
            "line": 1,
          },
          "start": {
            "col": 23,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`Parsers - arithmetic > should parse an arithmetic operation with in-between trailing spaces correctly 1`] = `
{
  "left": {
    "left": {
      "loc": {
        "end": {
          "col": 2,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "9",
    },
    "loc": {
      "end": {
        "col": 9,
        "line": 1,
      },
      "start": {
        "col": 1,
        "line": 1,
      },
    },
    "operator": "+",
    "right": {
      "loc": {
        "end": {
          "col": 9,
          "line": 1,
        },
        "start": {
          "col": 8,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "5",
    },
    "type": "BinaryExpression",
  },
  "loc": {
    "end": {
      "col": 36,
      "line": 1,
    },
    "start": {
      "col": 1,
      "line": 1,
    },
  },
  "operator": "-",
  "right": {
    "left": {
      "left": {
        "loc": {
          "end": {
            "col": 16,
            "line": 1,
          },
          "start": {
            "col": 15,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "loc": {
        "end": {
          "col": 27,
          "line": 1,
        },
        "start": {
          "col": 15,
          "line": 1,
        },
      },
      "operator": "*",
      "right": {
        "loc": {
          "end": {
            "col": 27,
            "line": 1,
          },
          "start": {
            "col": 26,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "4",
      },
      "type": "BinaryExpression",
    },
    "loc": {
      "end": {
        "col": 36,
        "line": 1,
      },
      "start": {
        "col": 15,
        "line": 1,
      },
    },
    "operator": "/",
    "right": {
      "loc": {
        "end": {
          "col": 36,
          "line": 1,
        },
        "start": {
          "col": 35,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "3",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`Parsers - arithmetic > should parse an arithmetic operation with trailing spaces at both ends correctly 1`] = `
{
  "left": {
    "left": {
      "loc": {
        "end": {
          "col": 5,
          "line": 1,
        },
        "start": {
          "col": 4,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "9",
    },
    "loc": {
      "end": {
        "col": 9,
        "line": 1,
      },
      "start": {
        "col": 4,
        "line": 1,
      },
    },
    "operator": "+",
    "right": {
      "loc": {
        "end": {
          "col": 9,
          "line": 1,
        },
        "start": {
          "col": 8,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "5",
    },
    "type": "BinaryExpression",
  },
  "loc": {
    "end": {
      "col": 27,
      "line": 1,
    },
    "start": {
      "col": 4,
      "line": 1,
    },
  },
  "operator": "-",
  "right": {
    "left": {
      "loc": {
        "end": {
          "col": 13,
          "line": 1,
        },
        "start": {
          "col": 12,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "4",
    },
    "loc": {
      "end": {
        "col": 27,
        "line": 1,
      },
      "start": {
        "col": 12,
        "line": 1,
      },
    },
    "operator": "*",
    "right": {
      "left": {
        "left": {
          "loc": {
            "end": {
              "col": 18,
              "line": 1,
            },
            "start": {
              "col": 17,
              "line": 1,
            },
          },
          "type": "NumberLiteral",
          "value": "4",
        },
        "loc": {
          "end": {
            "col": 22,
            "line": 1,
          },
          "start": {
            "col": 17,
            "line": 1,
          },
        },
        "operator": "/",
        "right": {
          "loc": {
            "end": {
              "col": 22,
              "line": 1,
            },
            "start": {
              "col": 21,
              "line": 1,
            },
          },
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "loc": {
        "end": {
          "col": 27,
          "line": 1,
        },
        "start": {
          "col": 16,
          "line": 1,
        },
      },
      "operator": "^",
      "right": {
        "loc": {
          "end": {
            "col": 27,
            "line": 1,
          },
          "start": {
            "col": 26,
            "line": 1,
          },
        },
        "type": "NumberLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;
